pkg/engine/engine.go

```diff
@@ -0,0 +1,88 @@
+package engine
+
+import (
+	"errors"
+
+	"github.com/klothoplatform/klotho/pkg/core"
+	knowledgebase "github.com/klothoplatform/klotho/pkg/knowledge_base"
+	"github.com/klothoplatform/klotho/pkg/provider"
+	"go.uber.org/zap"
+)
+
+type (
+	Engine struct {
+		Provider      provider.Provider
+		KnowledgeBase knowledgebase.EdgeKB
+	}
+)
+
+func NewEngine(provider provider.Provider) *Engine {
+	return &Engine{
+		Provider: provider,
+	}
+}
+
+func (e *Engine) Run(graph *core.ConstructGraph, constraints []Constraint, appName string) (*core.ResourceGraph, error) {
+
+	dag := core.NewResourceGraph()
+
+	for i := 1; i < 5; i++ {
+		zap.S().Infof("Running engine iteration %d", i)
+
+		// err := e.Provider.ExpandConstructs(graph, dag)
+		// if err != nil {
+		// 	return dag, err
+		// }
+
+		err := e.KnowledgeBase.ExpandEdges(dag, appName)
+		if err != nil {
+			return dag, err
+		}
+		validated, err := ValidateAndApply(dag, constraints)
+		if err != nil {
+			return dag, err
+		}
+		if validated {
+			break
+		}
+	}
+
+	var configurationErr error
+	for _, resource := range dag.ListResources() {
+		// Here we would apply the configuration based on node constraints we got
+		var configuration any
+		err := dag.CallConfigure(resource, configuration)
+		if err != nil {
+			errors.Join(configurationErr, err)
+		}
+	}
+	if configurationErr != nil {
+		return dag, configurationErr
+	}
+
+	err := e.KnowledgeBase.ConfigureFromEdgeData(dag)
+	if err != nil {
+		return dag, err
+	}
+
+	return dag, nil
+}
+
+func ValidateAndApply(graph *core.ResourceGraph, constraints []Constraint) (bool, error) {
+	var joinedErr error
+	allSatisfied := true
+	for _, constraint := range constraints {
```

Comment:
I think we should have a stable sorting of these — not just by level/scope (as the sort func above does), but even within each level/scope. I think a lot of constraints won't be commutative, so if we want "same graph + same constraints => same result" (which I think we should strive for), we need to make sure the constraints are applied in a consistent order.